### PR TITLE
feat(go): add comprehensive unit tests for command serialization

### LINKED ISSUE #2883: Go SDK: Add unit tests for command serialization/deserialization
## Description
The Go SDK contains numerous command types that implement the `MarshalBinary` interface to handle binary serialization. Currently, many of these implementations lack comprehensive unit tests, which is essential for ensuring data integrity and protocol compatibility.
    
## Task
Add unit tests to verify the correctness of the `MarshalBinary` implementations for all command types. The tests should ensure that the serialized output matches the expected binary format used by the Iggy server.
    
Go Command Types: [foreign/go/internal/command](https://github.com/apache/iggy/tree/master/foreign/go/internal/command). You should write tests for them.

### PR DESCRIPTION
**Pull request by saie-ch**
## Which issue does this PR close?

Closes #2883

## Rationale

The Go SDK contains numerous command types that implement the `MarshalBinary` interface for binary serialization, but many lacked comprehensive unit tests. Without thorough testing, serialization bugs can cause data corruption, runtime panics, and wire protocol incompatibility with the Iggy server and other SDKs.

## What changed?

**Before:** Go SDK had only 6 basic tests for command serialization with no edge case coverage.

   **After:** 88 unit tests (138 including subtests) covering all command types.

**Bugs Fixed:**
1. **CreatePersonalAccessToken** — `Expiry` field was `uint32` serialized with `PutUint32` at the wrong offset; changed to `uint64` with `PutUint64` to
   match the Rust wire format (`u64_le`)
   2. **CreateTopic / UpdateTopic** — `MarshalBinary` mutated the receiver by assigning `t.ReplicationFactor = new(uint8)` when nil; replaced with a local
   variable
   3. **UpdateUser** — `MarshalBinary` mutated the receiver by assigning `u.Username = new(string)` when nil; replaced with a local variable and fixed
   buffer size calculation

**Test Coverage:**
- Overall package: with 82.2% coverage. 
- All tests pass with the race detector enabled.

## What this PR does NOT fix

   - **Permissions.MarshalBinary() first byte corruption when streams are nil** — fixed in #3015

## Local Execution:

All 88 tests pass with race detector enabled and 82.2% statement coverage

## AI Usage

**Tool:** Claude Code (claude-sonnet-4.5)

**Scope:** Test implementation, bug discovery, and cross-SDK validation
- Helped design comprehensive test coverage strategy
- Generated test structures following Go best practices

### CODE REVIEW

File: foreign/go/internal/command/access_token_test.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+		Name:   "token",
+		Expiry: 0,
+	}
+
+	serialized, err := cmd.MarshalBinary()
+	if err != nil {
+		t.Fatalf("Failed to serialize CreatePersonalAccessToken with zero expiry: %v", err)
+	}
+
+	expected := []byte{
```
Comment (atharvalade) [inline foreign/go/internal/command/access_token_test.go:139]: These 4 zero bytes are not "Reserved/padding" -- the Rust wire format specifies `[expiry:u64_le]` starting at position 1 + `name_len` (see `core/binary_protocol/.../create_personal_access_token.rs:25`). The Go implementation uses `PutUint32` at `len(bytes)-4`, which writes the expiry in the high 32 bits of the `u64` field, causing the server to interpret the value as expiry << 32. The underlying bug is in `access_token.go:36` -- it should use `PutUint64(bytes[1+len(c.Name):]`, `uint64(c.Expiry))`, and the `Expiry` field type should be `uint64` to match the wire protocol. These tests currently codify the buggy serialization.

File: foreign/go/internal/command/user_test.go [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+		0x01,                   // Status = Active (1)
+		0x01,                   // Has permissions = 1
+		0x0B, 0x00, 0x00, 0x00, // Permissions length = 11
+		// Global permissions: 1,1,0,1,0,1,0,1,1,0
+		0x01, 0x01, 0x00, 0x01, 0x00, 0x01, 0x00, 0x01, 0x01, 0x00,
+		0x00, // No stream-specific permissions
+	}
+
+	if !bytes.Equal(serialized, expected) {
+		t.Errorf("Serialized bytes are incorrect.\nExpected:\t%v\nGot:\t\t%v", expected, serialized)
```
Comment (atharvalade) [inline foreign/go/internal/command/user_test.go:365]: This expected output is missing the `permissions_len:u32_le` field. The Rust wire format (see `create_user.rs:44-45`) always includes `permissions_len` on the wire, even when `has_permissions`=0. The expected bytes should end with 0x01, 0x00, 0x00, 0x00, 0x00, 0x00 (status, `has_permissions`=0, `permissions_len`=0). The underlying bug is in `user.go:77` where the nil-permissions branch only writes the flag byte but not the length field.

File: foreign/go/internal/command/user_test.go [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+		Id: userId,
+	}
+
+	serialized, err := cmd.MarshalBinary()
+	if err != nil {
+		t.Fatalf("Failed to serialize DeleteUser with string ID: %v", err)
+	}
+
+	expected := []byte{
+		0x02,                                                 // Kind = StringId
```
Comment (atharvalade) [inline foreign/go/internal/command/user_test.go:529]: This fix prevents the nil-permissions panic (good), but it's incomplete. The base length should be `len(userIdBytes)` + 1 + 4 because the Rust wire format always includes` permissions_len:u32_le` on the wire even when `has_permissions`=0 (see `update_permissions.rs:37-39`). As-is, the server will fail to decode the message because it reads 4 bytes for `permissions_len` that aren't present. The else branch at line 142 also needs to write `permissions_len`=0 (4 zero bytes) after the flag byte.

**Review by atharvalade — CHANGES_REQUESTED** [review]
overall looking good but I found a few things that need to be addressed

File: foreign/go/contracts/users.go
Comment (chengxilo) [inline foreign/go/contracts/users.go]: This part should not use byte(0) directly, doesn't align with the behavior in rust sdk.

https://github.com/apache/iggy/blob/9c8b5cc6ca682de6f3e7e0074224517e5488dc40/core/common/src/types/permissions/permissions_global.rs#L267-L273

    
  


@saie-ch would you mind to solve this in another PR first (to Implement unit test for permission and fix the problem in permission)?

File: foreign/go/contracts/users.go [hunk 4]
```diff
@@ -99,7 +99,7 @@ func (p *Permissions) MarshalBinary() ([]byte, error) {
 			}
 		}
 	} else {
-		bytes[0] = byte(0)
+		bytes[position] = byte(0)
```
Comment (chengxilo) [inline foreign/go/contracts/users.go:119]: You can solve this in the new PR too (sorry I thought it was a minor problem in the scope of this pr)

File: foreign/go/internal/command/update_user.go [hunk 5]
```diff
@@ -43,14 +43,14 @@ func (u *UpdateUser) MarshalBinary() ([]byte, error) {
 	username := *u.Username
```
Comment (chengxilo) [inline foreign/go/internal/command/update_user.go]: Here the `u.Username` is actually modified by the method, I don't think it's a good approach. Since you already modifed this method, would you mind to modified this too?

**Comment by atharvalade** [comment]
@saie-ch do you plan to implement anything else now that #3015 is merged? Once I get a green light from you, I can start review again and plan to merge this too after reviews. BTW I closed #3037 because of multiple conflicts and outdated code.

**Comment by saie-ch** [comment]
> @saie-ch do you plan to implement anything else now that #3015 is merged? Once I get a green light from you, I can start review again and plan to merge this too after reviews. BTW I closed #3037 because of multiple conflicts and outdated code.

Yes, please review. I believe, i've implemented the necessary changes.

**Comment by hubcio** [comment]
@saie-ch I noticed that comments from this PR are not fixed (output from claude code):
```
  #: 2958694591
  Reviewer: atharvalade
  Location: access_token.go:36
  Ask: Expiry field → uint64, use PutUint64 (current PutUint32 writes expiry into high 32 bits → server reads expiry << 32)
  Status: ❌ NOT fixed — file never touched. Line 24 still uint32, line 36 still PutUint32
  ────────────────────────────────────────
  #: 2958698450
  Reviewer: atharvalade
  Location: user.go:77 (CreateUser)
  Ask: nil-permissions branch must also write permissions_len:u32_le=0
  Status: ❌ NOT fixed — else branch still writes only 1 flag byte; capacity has no +4 for nil case
  ────────────────────────────────────────
  #: 2958701940
  Reviewer: atharvalade
  Location: user.go UpdatePermissions
  Ask: base length = len(userIdBytes)+1+4; else branch must write 4 zero bytes
  Status: ❌ NOT fixed — line 119 still len(userIdBytes)+1, else branch line 142 still single 0 byte. Code untouched
  ────────────────────────────────────────
  #: 2962372305
  Reviewer: chengxilo
  Location: update_user.go:43
  Ask: stop mutating receiver u.Username
  Status: ❌ NOT fixed — lines 39-43 still assign u.Username = new(string) then deref
  ```
  can you fix them?

/author

**Comment by saie-ch** [comment]
> @saie-ch I noticed that comments from this PR are not fixed (output from claude code):
> 
> ```
>   #: 2958694591
>   Reviewer: atharvalade
>   Location: access_token.go:36
>   Ask: Expiry field → uint64, use PutUint64 (current PutUint32 writes expiry into high 32 bits → server reads expiry << 32)
>   Status: ❌ NOT fixed — file never touched. Line 24 still uint32, line 36 still PutUint32
>   ────────────────────────────────────────
>   #: 2958698450
>   Reviewer: atharvalade
>   Location: user.go:77 (CreateUser)
>   Ask: nil-permissions branch must also write permissions_len:u32_le=0
>   Status: ❌ NOT fixed — else branch still writes only 1 flag byte; capacity has no +4 for nil case
>   ────────────────────────────────────────
>   #: 2958701940
>   Reviewer: atharvalade
>   Location: user.go UpdatePermissions
>   Ask: base length = len(userIdBytes)+1+4; else branch must write 4 zero bytes
>   Status: ❌ NOT fixed — line 119 still len(userIdBytes)+1, else branch line 142 still single 0 byte. Code untouched
>   ────────────────────────────────────────
>   #: 2962372305
>   Reviewer: chengxilo
>   Location: update_user.go:43
>   Ask: stop mutating receiver u.Username
>   Status: ❌ NOT fixed — lines 39-43 still assign u.Username = new(string) then deref
> ```
> 
> can you fix them?
> 
> /author

thanks @hubcio , will try to fix them.

**Comment by saie-ch** [comment]
@hubcio Fixed #1 (Expiry uint32→uint64) and #4 (stop mutating receiver). Also fixed the same mutation pattern in CreateTopic and UpdateTopic, and added non-mutation tests per @chengxilo's ask.                                                                                                                     
                                                                                                                                                               
 For #2 and #3 — the reviewer's suggestion to write permissions_len:u32_le=0 in the nil case is incorrect per the Rust reference. Both create_user.rs:63-64 and update_permissions.rs:53-54 only write put_u8(0) when permissions are None — no permissions_len field. The Rust test at create_user.rs:201 confirms: &[1,b'u', 1, b'p', 0, 0]. The real buffer bugs were already fixed by PR #3097. The Go code matches the Rust wire format as-is.

**Comment by hubcio** [comment]
you can use /ready at the beginning of the line to change label.

/ready

@saie-ch i will check this later in upcoming days.

File: foreign/go/internal/command/access_token_test.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	if !bytes.Equal(serialized, expected) {
+		t.Errorf("Serialized bytes are incorrect.\nExpected:\t%v\nGot:\t\t%v", expected, serialized)
+	}
+}
+
+// TestSerialize_CreatePersonalAccessToken_MaxExpiry tests edge case with maximum uint64 expiry
+func TestSerialize_CreatePersonalAccessToken_MaxExpiry(t *testing.T) {
+	cmd := CreatePersonalAccessToken{
+		Name:   "long_token",
+		Expiry: 18446744073709551615, // Max uint64 value
```
Comment (slbotbm) [inline foreign/go/internal/command/access_token_test.go]: This is a magic number. The following can be used instead:
```go
import "math"

x := uint64(math.MaxUint64)
```

File: foreign/go/internal/command/access_token_test.go
Comment (slbotbm) [inline foreign/go/internal/command/access_token_test.go:1]: It looks like `internal/command/access_token.go` does not have error paths for the commands, always returning `bytes, nil` or `[]bytes, nil`. Thus cases like `Name > 255 chars`  is not handled by the implementation and would not raise an error.

Comment (slbotbm) [reply foreign/go/internal/command/access_token_test.go:1]: Implementation could be a follow to this PR.

File: foreign/go/internal/command/consumer_group_test.go
Comment (slbotbm) [inline foreign/go/internal/command/consumer_group_test.go:1]: Same for this file as access_token.go. Validation for things like `Name > 255 chars` is not happening.

While `CreateConsumerGroup`, `GetConsumerGroups`, etc. return an `error`, there is no meaningful local validation / error paths.

File: foreign/go/internal/command/offset_test.go
Comment (slbotbm) [inline foreign/go/internal/command/offset_test.go:1]: Same as previously stated, no negative tests asserting MarshalBinary() errors.

File: foreign/go/internal/command/offset_test.go [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+
+// Helper function to create pointer to uint32
+func uint32Ptr(v uint32) *uint32 {
+	return &v
+}
+
+// TestSerialize_StoreConsumerOffsetRequest_WithPartition tests StoreConsumerOffset with partition specified
+func TestSerialize_StoreConsumerOffsetRequest_WithPartition(t *testing.T) {
+	consumerId, _ := iggcon.NewIdentifier(uint32(42))
+	streamId, _ := iggcon.NewIdentifier("stream1")
```
Comment (slbotbm) [inline foreign/go/internal/command/offset_test.go:36]: A test with `max uint32` can also be added for `StoreConsumerOffsetRequest`

File: foreign/go/internal/command/offset_test.go [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+	}
+
+	if !bytes.Equal(serialized, expected) {
+		t.Errorf("Serialized bytes are incorrect.\nExpected:\t%v\nGot:\t\t%v", expected, serialized)
+	}
+}
+
+// TestSerialize_GetConsumerOffset_WithoutPartition tests GetConsumerOffset without partition
+func TestSerialize_GetConsumerOffset_WithoutPartition(t *testing.T) {
+	consumerId, _ := iggcon.NewIdentifier(uint32(123))
```
Comment (slbotbm) [inline foreign/go/internal/command/offset_test.go:274]: another `max uint32` test for `GetConsumerOffset` as well

File: foreign/go/internal/command/access_token.go [hunk 9]
```diff
@@ -33,7 +33,7 @@ func (c *CreatePersonalAccessToken) MarshalBinary() ([]byte, error) {
 	bytes := make([]byte, length)
 	bytes[0] = byte(len(c.Name))
 	copy(bytes[1:], c.Name)
-	binary.LittleEndian.PutUint32(bytes[len(bytes)-4:], c.Expiry)
+	binary.LittleEndian.PutUint64(bytes[1+len(c.Name):], c.Expiry)
```
Comment (hubcio) [inline foreign/go/internal/command/access_token.go:36]: `copy(bytes[1:], c.Name)` writes into a dest of size `N+8` but only copies `N` bytes. works, but reads as if it could overflow. `copy(bytes[1:1+len(c.Name)], c.Name)` is clearer and lines up with the explicit `bytes[1+len(c.Name):]` index on the next line.

File: foreign/go/internal/command/update_user.go [hunk 10]
```diff
@@ -34,23 +34,22 @@ func (u *UpdateUser) MarshalBinary() ([]byte, error) {
...
+	if u.Username != nil {
+		username = *u.Username
 	}
 
-	username := *u.Username
-
 	if len(username) != 0 {
-		length += 2 + len(username)
+		length += 1 + len(username)
 	}
```
Comment (hubcio) [inline foreign/go/internal/command/update_user.go:49]: `len(username) == 0` collapses `Username: stringPtr("")` to the same wire bytes as `Username: nil` (has-username=0). that means a caller cannot distinguish "clear the username" from "leave it untouched" - the server sees the latter in both cases. `TestSerialize_UpdateUser_EmptyUsername` locks this behavior in. is the collapse intentional? if yes, leave a comment explaining it; if no, gate on the pointer (`u.Username != nil`) instead of the string length.

File: foreign/go/internal/command/session_test.go [hunk 11]
```diff
@@ -18,12 +18,48 @@
...
+	pos += 4
+	copy(buf[pos:], versionBytes)
+	pos += len(versionBytes)
+
+	binary.LittleEndian.PutUint32(buf[pos:], uint32(len(contextBytes)))
+	pos += 4
+	copy(buf[pos:], contextBytes)
+
+	return buf
+}
```
Comment (hubcio) [inline foreign/go/internal/command/session_test.go]: `buildExpectedLoginUser` rebuilds the expected bytes the same way the production code does (1-byte len prefix, `iggcon.Version`, same offsets). if prod is wrong, expected is wrong in lockstep and the assertion passes silently. consider either (a) hardcoding golden bytes for one fixed case with a known/substituted version, or (b) asserting each field individually against literal byte values instead of a mirror helper. the version case alone (`TestSerialize_LoginUser_ContainsVersion`) reads version from the same offset it was written at, so a wrong offset on both sides would still pass.

File: foreign/go/internal/command/marshal_nomutate_test.go [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+	username := "admin"
+	status := iggcon.Active
+	replicationFactor := uint8(1)
+
+	perms := &iggcon.Permissions{
+		Global: iggcon.GlobalPermissions{
+			ManageServers: true,
+			ReadServers:   true,
+		},
+	}
```
Comment (hubcio) [inline foreign/go/internal/command/marshal_nomutate_test.go]: `snap` here (and `&u` / `&s` / `&rf` / `perms` in the cases below) shares pointers with `cmd`. `reflect.DeepEqual` on the dereferenced struct values walks through pointers and compares the same backing memory, so a `MarshalBinary` that mutated `*Permissions` content (e.g. sorted `Streams` in place) or wrote through `*Username` would not be caught. the test only catches pointer reassignment - which is what the old `t.ReplicationFactor = new(uint8)` bug did, so it works for the fixes in this pr, but the name promises more than it delivers. deep-copy the receiver into `snapshot` (e.g. via a small helper that returns a value with freshly-allocated `*string`/`*uint8`/`*Permissions`) to actually verify no-mutation.

File: foreign/go/internal/command/partition_test.go
Comment (slbotbm) [inline foreign/go/internal/command/partition_test.go:1]: The code does not cover cases where `PartitionsCount = 0` is passed to `DeletePartitions`. Also `CreatePartitions` with max uint32 count.

Other than that, validation is missing from `foreign/go/internal/command/partition.go` as well

**Review by hubcio — CHANGES_REQUESTED** [review]
a few items that don't map to changed lines in this diff:

- `foreign/go/contracts/users.go` `Permissions.MarshalBinary` iterates `p.Streams` and `stream.Topics` (go maps) without sorting keys, so the produced wire bytes are non-deterministic across runs. two calls on the same `Permissions` value can produce different byte sequences. this file is not touched in this pr, but the new tests do not catch it either (since the structures used have either nil or single-entry maps). worth either fixing here (sort keys before iterating, e.g. `slices.Sorted(maps.Keys(p.Streams))`) or filing a follow-up.

- pr description says "Permissions.MarshalBinary() - Fixed first permission byte corruption when streams are nil" but `users.go` is unchanged in this pr. that fix landed earlier in #3015 (`169a09bcd`). consider rewording to "verified by new tests" or dropping the claim.

- pr description says "CreateUser - Fixed off-by-one buffer allocation (extra trailing byte)" but old `capacity = 4 + N + M` and new `1 + N + 1 + M + 1 + 1 = 4 + N + M` are arithmetically identical. the change is a readability refactor, not a bugfix.

File: foreign/go/internal/command/offset_test.go [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+func TestSerialize_DeleteConsumerOffset_MaxPartition(t *testing.T) {
+	consumerId, _ := iggcon.NewIdentifier(uint32(1))
+	streamId, _ := iggcon.NewIdentifier(uint32(2))
+	topicId, _ := iggcon.NewIdentifier(uint32(3))
+
+	cmd := DeleteConsumerOffset{
+		Consumer:    iggcon.NewSingleConsumer(consumerId),
+		StreamId:    streamId,
+		TopicId:     topicId,
+		PartitionId: uint32Ptr(4294967295), // Max uint32
```
Comment (slbotbm) [inline foreign/go/internal/command/offset_test.go]: 4294967295 is a magic number. Replace with 
```go
import "math"
x := uint32(math.MaxUint32)
```

File: foreign/go/internal/command/session_test.go
Comment (slbotbm) [inline foreign/go/internal/command/session_test.go:1]: Same pattern as the other files here - no negative/error-case tests here as well, and no validation in `session.go`

File: foreign/go/internal/command/system_test.go
Comment (slbotbm) [inline foreign/go/internal/command/system_test.go:1]: Same as others - no negative/error-case tests and no validation in `system.go`.

File: foreign/go/internal/command/system_test.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+	if !bytes.Equal(serialized, expected) {
+		t.Errorf("Serialized bytes are incorrect.\nExpected:\t%v\nGot:\t\t%v", expected, serialized)
+	}
+}
+
+// TestSerialize_GetClient_MaxValue tests serialization with maximum uint32 value (edge case)
+func TestSerialize_GetClient_MaxValue(t *testing.T) {
+	cmd := GetClient{
+		ClientID: 4294967295, // Max uint32 value
+	}
```
Comment (slbotbm) [inline foreign/go/internal/command/system_test.go]: Magic number

File: foreign/go/internal/command/update_user_test.go
Comment (slbotbm) [inline foreign/go/internal/command/update_user_test.go:1]: Same with this - no negative tests.

File: foreign/go/internal/command/user_test.go
Comment (slbotbm) [inline foreign/go/internal/command/user_test.go:1]: same - no negative tests.

**Review by slbotbm — CHANGES_REQUESTED** [review]
The changes look alright to me, though there is no validation for the input in most of the implementations that these tests test, and thus there are no negative tests. That means these will silently. This can be a follow-up PR.

Also, I noticed that there were tests testing serialization of empty strings. Aren't those supposed to be errors? Adding them shows that they are expected behaviour, not errors.

**Comment by saie-ch** [comment]
> The changes look alright to me, though there is no validation for the input in most of the implementations that these tests test, and thus there are no negative tests. That means these will silently. This can be a follow-up PR.
> 
> Also, I noticed that there were tests testing serialization of empty strings. Aren't those supposed to be errors? Adding them shows that they are expected behaviour, not errors.

The empty string tests reflect the current serialization behavior. Should we remove them until validation is added in the follow-up PR, as you said ?

**Comment by slbotbm** [comment]
> > The changes look alright to me, though there is no validation for the input in most of the implementations that these tests test, and thus there are no negative tests. That means these will silently. This can be a follow-up PR.
> > Also, I noticed that there were tests testing serialization of empty strings. Aren't those supposed to be errors? Adding them shows that they are expected behaviour, not errors.
> 
> The empty string tests reflect the current serialization behavior. Should we remove them until validation is added in the follow-up PR, as you said ?

maybe @chengxilo would be a better person to ask

**Comment by ryankert01** [comment]
I think we can change the title of this pr because it also somehow changes some code and fix bug?

**Comment by chengxilo** [comment]
> > > The changes look alright to me, though there is no validation for the input in most of the implementations that these tests test, and thus there are no negative tests. That means these will silently. This can be a follow-up PR.
> > > Also, I noticed that there were tests testing serialization of empty strings. Aren't those supposed to be errors? Adding them shows that they are expected behaviour, not errors.
> > 
> > 
> > The empty string tests reflect the current serialization behavior. Should we remove them until validation is added in the follow-up PR, as you said ?
> 
> maybe @chengxilo would be a better person to ask

If they are errors, I think it's better to just remove them. The tests should enforce correct behavior rather than just reflecting the current implementation. (I don't think I am a "better person" tho)

**Comment by hubcio** [comment]
Closing this due to lack of value added, no activity from author and code being too hard to review.
